### PR TITLE
Add a mock impl XKBWaylandKeyboardLayoutEngine

### PR DESCRIPTION
**Pull request by tonikitoo**
This is https://chromium-review.googlesource.com/c/chromium/src/+/849032.

[ozone/wayland] Add a mock impl XKBWaylandKeyboardLayoutEngine

In order to allow existing unittests (namely WaylandKeyboardTests)
and future ones to pass with XKB enabled, we need a specialization of
XkbWaylandKeyboardLayoutEngine for tests.
Following the logic used throughout the Wayland unittests' machinery,
it makes sense to mock a Xkb-capable KeyboardLayoutEngine, rather than
using an actual implementation.

This CL adds a mock implementation of XkbWaylandKeyboardLayoutEngine to
be used by Wayland unittests.

For this, it turns XkbWaylandKeyboardLayoutEngine into a pure virtual
class, which is implemented by

- XkbWaylandKeyboardLayoutEngineImpl, for regular ozone/wayland execution
(e.g. chrome/wayland), which keeps the current XkbWaylandKeyboardLayoutEngine
logic unchanged;

- MockXkbWaylandKeyboardLayoutEngine, a sibling implementation added
by this CL, derived from XkbWaylandKeyboardLayoutEngine to be used for
unittests.

BUG=578890
Issue #325

### CODE REVIEW

File: ui/ozone/platform/wayland/fake_server.cc [hunk 1]
```diff
@@ -15,6 +15,10 @@
 #include "base/run_loop.h"
 #include "base/strings/string_number_conversions.h"
 
+#if BUILDFLAG(USE_XKBCOMMON)
```
Comment (msisov) [inline ui/ozone/platform/wayland/fake_server.cc]: Maybe you better move the MockWaylandXkbKeyboardLayoutEngine to a separate source file and avoid adding checks for the USE_XKBCOMMON build flags all around here?

And all the mocked classes are inheritted from the ServerObject base class and kinda behave as server objects. MockWaylandXkbKeyboardLayoutEngine doesn't really suit this pattern much.

**Review by msisov — CHANGES_REQUESTED** [review]

File: ui/ozone/platform/wayland/fake_server.cc (hunk 1 above)
Comment (tonikitoo) [reply ui/ozone/platform/wayland/fake_server.cc]: that is a good point. Done.

**Review by msisov — DISMISSED** [review]

File: ui/ozone/platform/wayland/BUILD.gn [hunk 2]
```diff
@@ -113,5 +113,17 @@ source_set("wayland_unittests") {
     "//ui/ozone:platform",
   ]
 
+  import("//ui/base/ui_features.gni")
+  if (use_xkbcommon) {
+    sources += [
+      "mock_wayland_xkb_keyboard_layout_engine.cc",
```
Comment (msisov) [inline ui/ozone/platform/wayland/BUILD.gn:119]: you forgot to git add these classes, sir :) I can't see them

Comment (tonikitoo) [reply ui/ozone/platform/wayland/BUILD.gn:119]: Done.

File: ui/ozone/platform/wayland/mock_wayland_xkb_keyboard_layout_engine.cc [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ui/ozone/platform/wayland/mock_wayland_xkb_keyboard_layout_engine.h"
+#include "ui/events/keycodes/keyboard_code_conversion.h"
+
+#if !BUILDFLAG(USE_XKBCOMMON)
+#error "This file should only be compiled if USE_XKBCOMMON is enabled."
```
Comment (msisov) [inline ui/ozone/platform/wayland/mock_wayland_xkb_keyboard_layout_engine.cc:9]: nice!
 
nit: s/This file should/ This source code file can/ ???

File: ui/ozone/platform/wayland/mock_wayland_xkb_keyboard_layout_engine.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef UI_OZONE_PLATFORM_WAYLAND_MOCK_WAYLAND_XKB_KEYBOARD_LAYOUT_ENGINE_H_
+#define UI_OZONE_PLATFORM_WAYLAND_MOCK_WAYLAND_XKB_KEYBOARD_LAYOUT_ENGINE_H_
+
+#include "ui/base/ui_features.h"
+
+#if !BUILDFLAG(USE_XKBCOMMON)
+#error "This file should only be included if USE_XKBCOMMON is enabled."
```
Comment (msisov) [inline ui/ozone/platform/wayland/mock_wayland_xkb_keyboard_layout_engine.h:11]: the same nit as above

File: ui/ozone/platform/wayland/wayland_test.h [hunk 5]
```diff
@@ -7,11 +7,16 @@
 
 #include "base/message_loop/message_loop.h"
 #include "testing/gtest/include/gtest/gtest.h"
+#include "ui/base/ui_features.h"
```
Comment (msisov) [inline ui/ozone/platform/wayland/wayland_test.h:10]: is it still needed?

File: ui/ozone/platform/wayland/wayland_xkb_keyboard_layout_engine.cc [hunk 6]
```diff
@@ -23,10 +23,11 @@ void WaylandXkbKeyboardLayoutEngine::SetKeymap(xkb_keymap* keymap) {
   xkb_mod_indexes_.caps = xkb_keymap_mod_get_index(keymap, XKB_MOD_NAME_CAPS);
 }
 
-void WaylandXkbKeyboardLayoutEngine::UpdateModifiers(uint32_t depressed_mods,
-                                                     uint32_t latched_mods,
-                                                     uint32_t locked_mods,
-                                                     uint32_t group) {
+void WaylandXkbKeyboardLayoutEngineImpl::UpdateModifiers(
```
Comment (msisov) [inline ui/ozone/platform/wayland/wayland_xkb_keyboard_layout_engine.cc:26]: was it git cl format, which did it?

**Review by msisov — APPROVED** [review]
lgtm with nits

File: ui/ozone/platform/wayland/wayland_xkb_keyboard_layout_engine.cc (hunk 6 above)
Comment (tonikitoo) [reply ui/ozone/platform/wayland/wayland_xkb_keyboard_layout_engine.cc:26]: Right. it was git cl format's tweak.

File: ui/ozone/platform/wayland/wayland_test.h (hunk 5 above)
Comment (tonikitoo) [reply ui/ozone/platform/wayland/wayland_test.h:10]: it is needed to check #if BUILDFLAG(USE_XKBCOMMON) :/

**Pull request merged**